CMake changes for compatibility with Clang-cl for Windows, Qt 5.15

LouisStAmour opened the pull request:
See extensive changes to README, though we might want to split out setup instructions by platform into their own files. Also I've yet to fully regression test this with MSVC, GCC on Windows, or WSL on Linux.

LouisStAmour commented:
I'm testing Visual Studio 2019 now, and it uses a file named CMakeSettings.json to store the current settings which were manually configured in CLion. The equivalent CMakeSettings.json for Clang-cl.exe support in Visual Studio 2019 would be:

```json
{
  "configurations": [
    {
      "name": "x64-Debug",
      "generator": "Ninja",
      "configurationType": "Debug",
      "inheritEnvironments": [ "clang_cl_x64" ],
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "-v",
      "ctestCommandArgs": "",
      "variables": [
        {
          "name": "Boost_DEBUG",
          "value": "True",
          "type": "BOOL"
        },
        {
          "name": "BOOST_INCLUDEDIR",
          "value": "C:\\local\\boost_1_72_0",
          "type": "PATH"
        },
        {
          "name": "BOOST_LIBRARYDIR",
          "value": "C:\\local\\boost_1_72_0\\stage\\x64\\lib",
          "type": "PATH"
        },
        {
          "name": "BOOST_LIB_TOOLSET",
          "value": "vc142",
          "type": "STRING"
        },
        {
          "name": "Boost_COMPILER",
          "value": "-vc142",
          "type": "STRING"
        },
        {
          "name": "Qt5_DIR",
          "value": "C:\\Qt\\5.15.0\\msvc2019_64\\lib\\cmake\\Qt5",
          "type": "PATH"
        },
        {
          "name": "LLVM_DIR",
          "value": "C:\\llvm-project\\llvm\\build\\lib\\cmake\\llvm",
          "type": "PATH"
        },
        {
          "name": "Clang_DIR",
          "value": "C:\\llvm-project\\llvm\\build\\lib\\cmake\\clang",
          "type": "PATH"
        },
        {
          "name": "BUILD_CXX_LANGUAGE_PACKAGE",
          "value": "True",
          "type": "BOOL"
        }
      ],
      "cmakeExecutable": "C:\\Program Files\\CMake\\bin\\cmake.exe"
    }
  ]
}
```

egraether commented on CMakeLists.txt:
```diff
@@ -124,6 +138,7 @@ endif()
 
 if (BUILD_CXX_LANGUAGE_PACKAGE)
 
+	find_package(LLVM REQUIRED)
```
This should not be necessary. We used to have it, but removed it when revising this CMakeLists for open-sourcing. It should work when the environment variable `Clang_DIR` is set to the main `build` directory of LLVM (I think it is wrong in the README where it points to a subdirectory in the build directory).

egraether commented on CMakeLists.txt:
```diff
@@ -676,8 +693,7 @@ target_include_directories(${APP_PROJECT_NAME} SYSTEM
 # Use the Widgets module from Qt 5.
-qt5_use_modules(${APP_PROJECT_NAME} Widgets)
-qt5_use_modules(${APP_PROJECT_NAME} Network)
+target_link_libraries(${APP_PROJECT_NAME} Qt5::Network Qt5::Widgets)
```
This should also work properly when the environment variable `Qt5_DIR` is correctly set (or passed to cmake). Where does your `Qt5_DIR` point to?

LouisStAmour replied:
qt5_use_modules was flagged as deprecated, that’s why I changed this.

egraether replied:
Ah ok, thanks for clarification!

egraether commented on README.md:
```diff
@@ -122,6 +122,70 @@ If you are using the gui, we recommend that you activate advanced mode. Also you
+-DBOOST_LIBRARYDIR=C:\local\boost_1_72_0\stage\x64\lib
+-DBOOST_LIB_TOOLSET=vc142
+-DBoost_COMPILER=-vc142
+-DQt5_DIR=C:\Qt\5.15.0\msvc2019_64\lib\cmake\Qt5
```
Regarding the other comment: Try without `lib\cmake\Qt5`.

egraether commented on README.md:
```diff
@@ -122,6 +122,70 @@ If you are using the gui, we recommend that you activate advanced mode. Also you
+-DBoost_COMPILER=-vc142
+-DQt5_DIR=C:\Qt\5.15.0\msvc2019_64\lib\cmake\Qt5
+-DLLVM_DIR=C:\llvm-project\llvm\build\lib\cmake\llvm
+-DClang_DIR=C:\llvm-project\llvm\build\lib\cmake\clang
```
Regarding the other comment: Try without `lib\cmake\clang`.

egraether commented on README.md:
```diff
@@ -122,6 +122,70 @@ If you are using the gui, we recommend that you activate advanced mode. Also you
+-DBOOST_LIB_TOOLSET=vc142
+-DBoost_COMPILER=-vc142
+-DQt5_DIR=C:\Qt\5.15.0\msvc2019_64\lib\cmake\Qt5
+-DLLVM_DIR=C:\llvm-project\llvm\build\lib\cmake\llvm
```
Regarding the other comment: Remove this line.

egraether commented on README.md:
```diff
@@ -122,6 +122,70 @@ If you are using the gui, we recommend that you activate advanced mode. Also you
+
+* [Build Tools for Visual Studio 2019](https://aka.ms/buildtools) (as an alternative to installing Visual Studio)
+* [Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk/) for released versions of Windows. If running a Windows Insider preview, you'll need the [Preview SDK](https://insider.windows.com/en-us/for-developers/) instead.
+* Boost 1.72.0 compiled [following Windows instructions](https://www.boost.org/doc/libs/1_72_0/more/getting_started/windows.html) from within [a Build Tools command prompt](https://docs.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=vs-2019#developer_command_file_locations) (or you can download a [pre-compiled set of binaries](https://bintray.com/boostorg/release/boost-binaries)) ...
```
We currently use Boost 1.67, which is the version used in certain Linux distributions. We currently don't plan to update this to a higher version, unless there is a good reason.

LouisStAmour replied:
1.67 is two days shy of being two years old. While the changelogs don’t reveal anything especially interesting since then, they do mention adding private visibility for smaller shared libraries, adding more C++17 feature compatibility and so on.

Boost, like Qt, might be the type of library where we need CI runs with more than one version: the minimum we support, and then the maximum we support. Similar to Mac, Windows and Linux compatibility, we could define platform to include these different variants. The easiest way to do that might be to define a set of Linux OS releases to support and test compiling and linking on those OSes.

LouisStAmour replied:
It wouldn’t be perfect (kernel, etc.) but using Docker we can mimic the filesystem and dependencies shipped by various operating systems, and list those OSes as platforms we support. My preference, though it is just a preference, would be for us to always support the latest version, and introduce at most what I would call temporary hacks or workarounds to support previous versions. Boost is a stable library, it’s easy to support multiple versions. Same with Qt if you’re not trying to use the latest and greatest. Clang on the other hand seems incredibly unstable, so we probably can’t support back farther than one, maybe two, releases, and new releases seem to occur every 6 months. Java on the other hand seems relatively stable but again you can only use features from the oldest version of Java you’d want to support...

egraether replied:
I generally support updating software to the newest version. However, regarding Boost there is really not much of a benefit for us in doing so. And testing with multiple Boost versions is not a problem worth solving in my opinion.

Our officially supported version is `1.67` and everything in our repository should consistently reflect this choice. If you use a different version that also works, that is your personal choice. When there is a good reason to update, then we will move forward.

egraether commented on README.md:
```diff
@@ -122,6 +122,70 @@ If you are using the gui, we recommend that you activate advanced mode. Also you
 
 After generating the build configuration, just open the Sourcetrail.sln file that was generated by CMake and build the Sourcetrail project.
 
+##### If using clang-cl.exe and Ninja (such as via CLion)
```
This is interesting information, but I'm not sure we want it here. @mlangkabel wanted to revise the README. What do you think?

LouisStAmour replied:
My preference would be to break the info in the readme into multiple files, perhaps re-organized by platform.

egraether replied:
Sounds good!

mlangkabel replied:
@LouisStAmour, thanks for providing setup instructions for using Clang/Ninja on Windows. Currently the document was structured this way:

* target 1
  * dependencies
    * dependency 1
    * ...
    * dependency N (windows only)
    * dependency M (unix only)
  * build instructions
    * instruction windows
    * instruction unix
* target 2
  * ...
We chose this approach because it allows us to keep all (cross-platform) information for a target together in one spot. So if we update a dependency version we don't have 3 documents where we need to remember to make the update. And that's why I would like to keep it that way. 

However, the format of this whole "Building Sourcetrail" section is still horrible to read. I was working on some improvements. Maybe I can get that done today. This will probably cause conflicts here, but you would need to split the information provided here into the individual sections (e.g. dependencies, building).

LouisStAmour commented:
Haven't yet addressed the above comments, did find another tweak for compatibility to apply. I've been having trouble with an error message from boost interprocess when Clang is emulating MSVC however.... not sure if there's an easy fix or if I'll have to find a way to switch to another compiler.
